File: disjoint_set.cpp
/*
A disjoint set is a data structure for finding if elements belong to the same or different groups

Find and Union can be performed on it
*/
#include <bits/stdc++.h>

class disjoint_set{
public:
	std::map<int, int> set_map;
	std::vector<std::pair<int, int>> elist;

	// default constructor
	disjoint_set(std::vector<std::pair<int,int>> edge_list){
		for(int i = 0; i < edge_list.end()-edge_list.begin(); i++){
			set_map[edge_list[i].first] = -1;
			set_map[edge_list[i].second] = -1;
		}

		elist = edge_list;
	}
public:
	bool detect_loop(){
		int temp,temp2;
		for(int i = 0; i < elist.end()-elist.begin(); i++){
			int par_u = find_parent(elist[i].first), par_v = find_parent(elist[i].second);
			if (par_u == par_v){
				return true;
			}
			else{
				// Make sure that the following three condidtions are taken care of
				if(set_map[elist[i].first] != -1)
					set_map[elist[i].first] = find_parent(elist[i].second);
				else if(set_map[elist[i].second] != -1)
					set_map[elist[i].second] = find_parent(elist[i].first);
				else
					temp = elist[i].second;
					while(set_map[temp]!= -1){
						temp2 = set_map[temp];
						set_map[temp] = find_parent(elist[i].first);
						temp = temp2;
					}
					set_map[temp] = find_parent(elist[i].first);
			}
		}

		return false;
	}

	int find_parent(int vertex){
		int par = vertex;
		while(set_map[par] != -1){
			par = set_map[par] ;
		}
		return par;
	}
};

int main(){

	std::vector<std::pair<int,int>> graph = {{7, 6}, {8, 2}, {6, 5}, {0, 1}, {2, 5}, {2, 3}, {0, 7}, {1, 2}};
	// std:: cout << "compiles";
	disjoint_set l(graph);

	if (l.detect_loop()){
		std::cout << "loop detected\n";
	}
	else{
		std::cout << "no loop\n";
	}
	return 0;
}

File: disjoint_set.py
'''
A disjoint set is a data structure for finding if elements belong to the same or different groups

Find and Union can be performed on it
'''

class disjoint_set:
	set_map = {}
	edge_list = []
	def detect_loop(self, edge_list):
		# initialize the edge list
		self.edge_list = edge_list
		for edge in edge_list:
			self.set_map[edge[0]], self.set_map[edge[1]] = -1, -1

		for edge in self.edge_list:
			par_u, par_v = self.find_parent(edge[0]), self.find_parent(edge[1])
			
			print(edge[0], edge[1])
			print("\n==\n", self.set_map, "\n==\n")
			if par_u == par_v:
				return True
			else:
				# This distiction is very important. Improper implementation of
				# this part will really cost you!

				# Added another important while loop. Now the algorithm is correct. God damn
				if self.set_map[edge[0]] == -1:
					self.set_map[edge[0]] = self.find_parent(edge[1])#edge[1]
				elif self.set_map[edge[1]] == -1:
					self.set_map[edge[1]] = self.find_parent(edge[0])#edge[1]
				else:
					temp = edge[1]
					while self.set_map[temp] != -1:
						temp2 = self.set_map[temp]
						self.set_map[temp] = self.find_parent(edge[0])
						temp = temp2
					self.set_map[temp] = self.find_parent(edge[0])

				#self.union(self.find_parent(par_u), self.find_parent(par_v))
		return False

	def union(self, u, v):
		set_map[u] = v

	def find_parent(self, vertex):
		par = vertex
		# self.set_map[par]
		while(self.set_map[par] != -1):
			par = self.set_map[par]
		return par

graph =  [(7, 6), (8, 2), (6, 5), (0, 1), (2, 5), (2, 3), (0, 7), (1, 2)]

l = disjoint_set()
print(l.detect_loop(graph))